lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystem.java

```diff
@@ -202,4 +211,34 @@ private static void validateS3Location(Location location)
     {
         new S3Location(location);
     }
+
+    private boolean shouldReadObject(S3Object object)
```

Comment:
NIT: Maybe move connected methods to `S3ObjectStorageClassFilter` and return function from context
```
.filter(context.s3ObjectStorageClassFilter())
```
e.g
```
public enum S3ObjectStorageClassFilter
    {
        READ_ALL(o -> true),
        SKIP_ALL_GLACIER(S3ObjectStorageClassFilter::isNotGlacierObject),
        READ_RESTORED_GLACIER_OBJECTS(S3ObjectStorageClassFilter::isCompletedRestoredObject);
        
        private Function<S3Object, Boolean> filter;

       private S3ObjectStorageClassFilter(Function<S3Object, Boolean> filter){
          .....
       }

    private static boolean isCompletedRestoredObject(S3Object object)
    {
        /* There are 3 cases for the restore status:
         *
         * 1. The object is not restored, and has not been requested to be restored. We should have a null
         * value for restoreStatus
         * 2. The object is in the process of being restored. isRestoreInProgress will be true, but restoreExpiryDate
         * will be null.
         * 3. The object has completed restore. isRestoreInProgress will be false, restoreExpiryDate will be some date.
         *
         * Since we only care about distinguishing when it is case 3, we can just check restoreExpiryDate for a non-null value. */
        return isNotGlacierObject(object) || Optional.ofNullable(object.restoreStatus())
                .map(RestoreStatus::restoreExpiryDate)
                .isPresent();
    }

    private static boolean isNotGlacierObject(S3Object object)
    {
        return !GLACIER_STORAGE_CLASSES.contains(object.storageClass());
    }
    }
```

Reply:
I like this! Will update